pkgs/common-updater/unstable-updater.nix

```diff
@@ -78,27 +88,51 @@ let
     pushd "$tmpdir"
     commit_date="$(${git}/bin/git show -s --pretty='format:%cs')"
     commit_sha="$(${git}/bin/git show -s --pretty='format:%H')"
-    if [[ -z "$use_stable_version" ]]; then
-        new_version="unstable-$commit_date"
-    else
-        depth=100
-        while (( $depth < 10000 )); do
-            last_tag="$(${git}/bin/git describe --tags --abbrev=0 2> /dev/null || true)"
-            if [[ -n "$last_tag" ]]; then
-                break
+    last_tag=""
+    if [[ -z "$hardcode_zero_version" ]]; then
+        if [[ "$shallow_clone" == "1" ]]; then
+            depth=100
+            while (( $depth < 10000 )); do
+                last_tag="$(${git}/bin/git describe --tags --abbrev=0 2> /dev/null || true)"
+                if [[ -n "$last_tag" ]]; then
+                    tag_matched="$(echo ''${last_tag} | grep -Eo ''${tag_format} || true)"
+                    if [[ -z "$tag_matched" ]]; then
+                        ${lib.getExe git} tag -d ''${last_tag}
+                        continue
+                    fi
+                    break
+                fi
+                ${git}/bin/git fetch --depth="$depth" --tags
+                depth=$(( $depth * 2 ))
+            done
+
+            if [[ -z "$last_tag" ]]; then
+                # To be extra sure, check if full history helps with finding a tag
+                git fetch --tags
```

Comment:
`git` isn't on PATH, is it? would this work?

Maybe we should convert this to `writeShellApplication`, so we can use `runtimeInputs` and not have to repeat `${git}/bin/git` or `${lib.getExe git}` everywhere - and get the `shellcheck` benefits of that builder.

Reply:
Refactored it to use `writeShellApplication`, added used packages to `runtimeInputs` and fixed as much as I could about the shellcheck messages.